text/0060-bazel.md

```diff
@@ -0,0 +1,137 @@
+---
+feature name: bazel
+start date: 2020-01-16
+rfc pr: (leave this empty)
+related issue: 0060
+---
+
+# Summary
+
+This proposal is for migrating the build system to [Bazel](https://bazel.build/), as it solves
+many if not all of the deficiencies of the current system. This involves adding Bazel resources and rules to the repository to manage
+build and test workflows, and bringing the team up to speed on how to use them.
+
+# Motivation
+
+The CDK currently uses a combination of [Lerna](https://lerna.js.org/), [Yarn Workspaces](https://yarnpkg.com/lang/en/docs/workspaces/), 
+and Bash scripts to orchestrate build and test workflows. This has several advantages for managing a monorepo structure, but can also
+be very slow, difficult to work with and extend, and hard to parallelize amongst the team (ie when one unrelated resource changes, the
+whole project still needs to be rebuilt). 
+
+# Basic Example
+
+Instead of running `bash build.sh` or `npm run build` on a subpackage, an engineer would simply run `bazel build <path-to-resource>` or
+`bazel test <path-to-resource`. Running a specific Bazel rule (for instance orchestrating a release) is also straightforward, with
+`bazel run <path-to-resource>.<action>`.
+
+# Design Summary
+
+In order to setup Bazel, the Bazel-specific rules and resources need to be added to the repository. This means adding `BUILD.bazel` files
+to each subpackage, and then building custom rules around some of the internals of the project (e.g. CFN to TS).
+
+# Detailed Design
+
+The structure of adding Bazel to a project is quite simple. Represented below is a tree of the files
+needed for a standard project
+
+```
+BAZEL.build
+package.json
+|packages|
+  BUILD.bazel
+  |apigateway|
+    BUILD.bazel
+    package.json
+    apigateway.ts
+    index.ts
+WORKSPACE <-- Bazel project definition, only needed once
+yarn.lock
+```
+
+The `BUILD` files themselves are quite simple, since Bazel has rules built-in for compiling a TypeScript library and for
+running JavaScript tests. We would need to write custom rules in the 
+[Skylark language](https://docs.bazel.build/versions/master/skylark/language.html) to perform custom operations like
+building TypeScript definitions from CloudFormation definition files. This is relatively straightforward since the tooling
+for this exists in TypeScript already, and Skylark is a subset of Python.
+
+Example `BUILD.bazel` file, assuming `cdk_library` is a Bazel rule that calls `ts_library` with another rule that performs 
+TS transformation:
+```
+load("//tools:defaults.bzl", "cdk_library", "pkg_npm")
+
+package(default_visibility = ["//visibility:public"])
+
+cdk_library(
+    name = "aws-amazonmq",
+    scope = ["AWS::AmazonMQ"],
+    outputs = ["lib/amazonmq.generated.ts"],
+    srcs = glob(["lib/**/*.ts"]),
+    deps = [
+        "//packages/@aws-cdk/cdk",
```

Comment:
It feels like this information is duplicated from the `package.json` file, right? When I investigated `bazel` a  short while ago this made me feel that those `BUILD.bazel` file should be generated from existing project meta-information... Duplicating dependency declarations is neither enjoyable nor safe in the long run.

Reply:
As mentioned in Elad's comment above, while this would be duplicated for now, it would not be duplicated in the long run as the Bazel team is working on auto-extraction of dependencies from package.json files. However, I think it's worth biting the bullet for now, because it is just the short term, and the risks are minimal in my opinion. For instance, when was the last time you updated the dependencies of a long-standing CDK package? And even if you did, if you skipped adding it to the Bazel file, it wouldn't build, and if you skipped adding it to the package.json file, the integration tests wouldn't run.